src/lib/autocontainers.ts

```diff
@@ -0,0 +1,175 @@
+        // Do not handle private tabs or invalid tabIds.
+        if (details.tabId === -1) return
+        let tab = await browser.tabs.get(details.tabId)
+        if (tab.incognito) return
+
+        // Only handle http requests.
+        if (details.url.search("^https?://") < 0) return
+
+        let cookieStoreId = await this.parseAucons(details)
+
+        // Silently return if we're already in the correct container.
+        if (tab.cookieStoreId === cookieStoreId) return
+
+        if (this.cancelEarly(tab, details)) return { cancel: true }
+
+        browser.tabs
+            .create({
+                url: details.url,
+                cookieStoreId: cookieStoreId,
+                active: tab.active,
+                windowId: tab.windowId,
+            })
+            .then(_ => {
+                if (details.originUrl) {
+                    window.history.back()
+                } else {
+                    browser.tabs.remove(details.tabId)
+                }
+            })
+
+        return { cancel: true }
+    }
+
+    //Handles the requests after the initial checks made in this.autoContain.
+    cancelEarly = (tab: browser.tabs.Tab, details: IDetails): boolean => {
+        if (!this.cancelledRequests[tab.id]) {
+            this.cancelRequest(tab, details)
+        } else {
+            let cancel = false
+            let cr = this.getCancelledRequest(tab.id)
+
+            if (cr.requestIds[details.requestId] || cr.urls[details.url]) {
+                cancel = true
+            }
+
+            cr.requestIds[details.requestId] = true
+            cr.urls[details.url] = true
+
+            return cancel
+        }
+        return false
+    }
+
+    cancelRequest = (tab: browser.tabs.Tab, details: IDetails): void => {
+        this.cancelledRequests[tab.id] = {
+            requestIds: {
+                [details.requestId]: true,
+            },
+            urls: {
+                [details.url]: true,
+            },
+        }
+
+        // The webRequest events onCompleted and onErrorOccurred are not 100% reliable.
+        // Mozilla's contain facebook extension points this out and adds a guaranteed removal so we'll do the same.
+        setTimeout(() => {
+            this.clearCancelledRequests(tab.id)
+        }, 2000)
+    }
+
+    getCancelledRequest = (tabId: number): ICancelledRequest => {
+        return this.cancelledRequests[tabId]
+    }
+
+    // Clear the cancelled requests.
+    clearCancelledRequests = (tabId: number): void => {
+        if (this.cancelledRequests[tabId]) {
+            delete this.cancelledRequests[tabId]
+        }
+    }
+
+    // Parses autocontain directives and returns valid cookieStoreIds or errors.
+    parseAucons = async (details): Promise<string> => {
+        let aucons = Config.get("autocontain")
+        const ausites = Object.keys(aucons)
+        const aukeyarr = ausites.filter(
+            e => details.url.search("^https?://.*" + e + "/") >= 0,
```

Comment:
I think the `autocontain` documentation is slightly misleading. It isn't clear that your "URL" is actually parsed as a regex.

E.g, `autocontain www.google.co.uk Work` would also match on `www#google#co.uk`and maybe `"https://www.duckduckgo.com/q=?www.google.co.uk/".search("^https?://.*" + "www.google.co.uk" + "/")`

It's also a useful property as it means you can do e.g. `autocontain www.google.co(.uk|com)`. (Forgive the pidgin regex).

Reply:
I agree, I tried to make it clear by waffling on about domains. I'll write something to absolutely make it clear.

The regex bit is indeed useful and should probably also be documented better. 

I had already encountered similar situations as: `"https://www.duckduckgo.com/q=?www.google.co.uk/".search("^https?://.*" + "www.google.co.uk" + "/")` and had made some changes to the matching. I guess I could take another stab at writing a better regex there.